server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -54,134 +56,204 @@
-            this.previousMetaData = null;
-            previouslyWrittenIndices = emptySet();
+        if (event.state().blocks().disableStatePersistence()) {
+            // reset the current state, we need to start fresh...
+            previousMetaData = null;
+            previousManifest = null;
             return;
         }
 
-        MetaData newMetaData = state.metaData();
-        // we don't check if metaData changed, since we might be called several times and we need to check dangling...
-        Set<Index> relevantIndices = Collections.emptySet();
-        boolean success = true;
-        // write the state if this node is a master eligible node or if it is a data node and has shards allocated on it
-        if (state.nodes().getLocalNode().isMasterNode() || state.nodes().getLocalNode().isDataNode()) {
-            if (previousMetaData == null) {
-                try {
-                    // we determine if or if not we write meta data on data only nodes by looking at the shard routing
-                    // and only write if a shard of this index is allocated on this node
-                    // however, closed indices do not appear in the shard routing. if the meta data for a closed index is
-                    // updated it will therefore not be written in case the list of previouslyWrittenIndices is empty (because state
-                    // persistence was disabled or the node was restarted), see getRelevantIndicesOnDataOnlyNode().
-                    // we therefore have to check here if we have shards on disk and add their indices to the previouslyWrittenIndices list
-                    if (isDataOnlyNode(state)) {
-                        Set<Index> newPreviouslyWrittenIndices = new HashSet<>(previouslyWrittenIndices.size());
-                        for (IndexMetaData indexMetaData : newMetaData) {
-                            IndexMetaData indexMetaDataOnDisk = null;
-                            if (indexMetaData.getState().equals(IndexMetaData.State.CLOSE)) {
-                                indexMetaDataOnDisk = metaStateService.loadIndexState(indexMetaData.getIndex());
-                            }
-                            if (indexMetaDataOnDisk != null) {
-                                newPreviouslyWrittenIndices.add(indexMetaDataOnDisk.getIndex());
-                            }
-                        }
-                        newPreviouslyWrittenIndices.addAll(previouslyWrittenIndices);
-                        previouslyWrittenIndices = unmodifiableSet(newPreviouslyWrittenIndices);
-                    }
-                } catch (Exception e) {
-                    success = false;
-                }
+        try {
+            // if previousManifest is null we load it from disk
+            // we don't do the same for previous metadata, because our comparison can show that nothing was changed,
+            // but in fact it has changed
```

Comment:
1) We load the previous manifest because we need it to access get the set of previously written indices
2) We don't load the previous metadata, because we always want to re-write metadata when the node joins the cluster. Remember you said that, for example, our IndexMetaData comparison can show that IndexMetaData is not changed, but in fact, it has changed.
Please let me know if it's clear for you or should I re-phrase the comment.

Reply:
Ok, I've removed the comment and the line loading manifest, because you suggested to load manifest only on data node if we really need it.